### PR TITLE
Add local log - exp optimization

### PR DESCRIPTION
**Pull request by ricardoV94**
It seems there were no automatic rewrites to remove `log(exp(x))` or `exp(log(x))` expressions. Or at least I couldn't get them to show up when playing with optimizations.

Stealing some of @brandonwillard code from the Discourse:
```python
def optimize_graphs(*graphs, verbose=False):
    inputs = list(at_inputs(graphs))
    graphs = list(graphs)
    fgraph = FunctionGraph(inputs, graphs, clone=False)
    canonicalize_opt = optdb.query(Query(include=['fast_run']))
    with config.change_flags(optimizer_verbose=verbose):
        canonicalize_opt(fgraph)
    print(aesara.pp(fgraph.outputs[0]))

x = at.vector('x')
optimize_graphs(at.log(at.exp(x)))  # x
optimize_graphs(at.exp(at.log(x)))  # x
```
While in master it prints:
```python
optimize_graphs(at.log(at.exp(x)))  # Elemwise{Composite{log(exp(i0))}}[(0, 0)](x)
optimize_graphs(at.exp(at.log(x)))  # Elemwise{Composite{exp(log(i0))}}[(0, 0)](x)
```

Would it be as simple as this? Obviously it would require tests, but I wanted to first check whether I am missing something obvious... and also check how many tests this breaks.
***

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [x] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [x] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [x] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.  If your commit description starts with "Fix...", then you're probably making this mistake.
+ [x] There are tests covering the changes introduced in the PR.

Don't worry, your PR doesn't need to be in perfect order to submit it.  As development progresses and/or reviewers request changes, you can always [rewrite the history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) of your feature/PR branches.

If your PR is an ongoing effort and you would like to involve us in the process, simply make it a [draft PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests).

### CODE REVIEW

**Comment by brandonwillard** [comment]
The latter optimization, `at.exp(at.log(x))`, might need to produce something like `at.switch(at.ge(x, 0), x, np.nan)`.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
The latter optimization, `at.exp(at.log(x))`, might need to produce something like `at.switch(at.ge(x, 0), x, np.nan)`.

**Comment by ricardoV94** [comment]
> The latter optimization, `at.exp(at.log(x))`, might need to produce something like `at.switch(at.ge(x, 0), x, np.nan)`.

Good point, I knew it couldn't be so simple! 

Is it still worthwhile in that case or should we just go with the first one?

**Comment by ricardoV94** [comment]
The `log(exp(x))` rewrite gives a sizeable speedup as expected of ~4 orders of magnitude, but the `exp(log(x))` with the appropriate `switch` is only ~3x faster than the original. Perhaps the second does not justify the added graph complexity?

**Comment by brandonwillard** [comment]
> The `log(exp(x))` rewrite gives a sizeable speedup of ~4 orders of magnitude, but the `exp(log(x))` with the appropriate `switch` is only ~3x faster than the original. Perhaps the second does not justify the added graph complexity?

Can you post those comparisons?

**Comment by ricardoV94** [comment]
> Can you post those comparisons?

Actually I forgot I was testing everything with C-compilation disabled. Here is what I did with the latest pushed changes:

```python
#%%
# Confirm optimizations are enabled/ disabled when testing performance
x = at.vector('x')
xs = np.linspace(-5000, 5000, 10_000, dtype='int32')

optimize_graphs(at.log(at.exp(x)))
optimize_graphs(at.exp(at.log(x)))

#%%
# Compile and run log(exp(x))
f = aesara.function([x], at.log(at.exp(x)))
print(f(xs))

#%%
%timeit f(xs)

#%%
# Compile and run (exp(log(x))
f = aesara.function([x], at.exp(at.log(x)))
print(f(xs))

#%%
%timeit f(xs)
```

Optimization | log(exp(x)) | exp(log(x))
--- | --- | ---
On | 33 µs | 43 µs
Off | 207 µs | 166 µs

Now both are actually comparable, and about ~3-6x faster

File: aesara/tensor/math_opt.py [hunk 1]
```diff
@@ -226,6 +227,42 @@ def local_func_inv(fgraph, node):
...
+    if isinstance(prev_op, aes.Exp) and isinstance(node_op, aes.Log):
+        return x.owner.inputs
+
+    if isinstance(prev_op, aes.Log) and isinstance(node_op, aes.Exp):
+        out = node.outputs[0]
+
+        x = x.owner.inputs[0]
+        new_out = switch(ge(x, 0), x, np.nan)
+
+        if new_out.dtype != out.dtype:
```
Comment (ricardoV94) [inline aesara/tensor/math_opt.py:258]: Not sure if this is needed. Does switch ever change the dtype of it's inputs?

Comment (brandonwillard) [reply aesara/tensor/math_opt.py:258]: I don't believe there is a 32bit NaN, so that should always upcast to `float64`.

Comment (ricardoV94) [reply aesara/tensor/math_opt.py:258]: float32 `nan` seems fine:
```python
x = at.fvector('x')
y = at.cast(at.switch(x > 0, x, np.nan), 'float32')
f = aesara.function([x], y)

print( f([-5.0]), f([-5.0]).dtype)
# [nan] float32
```

**Comment by twiecki** [comment]
Wow, this could be massive. Can't believe this optimization didn't exist already.

**Comment by ricardoV94** [comment]
The new test for exp(log(x)) is failing here, but passing locally. One of the conditions the original tests were checking was that the length of `f.maker.fgraph.toposort()` is 1 for optimized functions and either 1 or 2 for functions that should not be optimized.

Locally I get a length of 1 for the failing case 
```python
[Elemwise{Composite{Cast{float32}(Switch(GE(i0, i1), i0, i2))}}(<TensorType(float32, matrix)>, TensorConstant{(1, 1) of 0}, TensorConstant{[[nan]]})]
```

But here I get 3:
```python
[
    Elemwise{ge,no_inplace}(<TensorType(float32, matrix)>, TensorConstant{(1, 1) of 0}), 
    Elemwise{switch,no_inplace}(Elemwise{ge,no_inplace}.0, <TensorType(float32, matrix)>, TensorConstant{[[nan]]}), 
    Elemwise{Cast{float32}}(Elemwise{switch,no_inplace}.0)
]
```
Any idea why I am not getting this locally?

**Comment by ricardoV94** [comment]
> Wow, this could be massive. Can't believe this optimization didn't exist already.

We might also want to add the equivalent rewrites for the custom `log1p` and `log1pexp` (i.e., softplus):  so that
* `at.exp(at.log1p(x)) = 1 + x`
* `at.exp(at.nnet.softplus(x)) = 1 + exp(x)`

File: aesara/tensor/math_opt.py (hunk 1 above)
Comment (brandonwillard) [reply aesara/tensor/math_opt.py]: As long as there's a usable 16/32 `nan`, then I'm fine with the cast to the original `dtype`.

File: aesara/tensor/math_opt.py [hunk 2]
```diff
@@ -219,6 +221,19 @@ def local_func_inv(fgraph, node):
...
+            if isinstance(prev_op, aes.Log):
+                x = x.owner.inputs[0]
+                new_out = switch(ge(x, 0), x, np.nan)
+
+                old_out = node.outputs[0]
+                if new_out.dtype != old_out.dtype:
+                    new_out = cast(new_out, dtype=old_out.dtype)
+                if new_out.type != old_out.type:
+                    return
+                return [new_out]
```
Comment (brandonwillard) [inline aesara/tensor/math_opt.py:235]: This is supposed to be a generalized implementation, so let's not hard-code special cases.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/tensor/math_opt.py (hunk 2 above)
Comment (ricardoV94) [reply aesara/tensor/math_opt.py:235]: I see the point, I was trying to avoid checking ever pair of nodes twice in different functions.

**Comment by ricardoV94** [comment]
Alright, so I separated the new opts (and tests) into a separate function again. I also added an opt for `exp(log1p(x))` as this shares most of the logic with the `exp(log(x))` rewrite. 

It also seems I don't need an explicit casting of the switch output. It's enough if I convert `nan` to the original output dtype inside the switch.

**Comment by ricardoV94** [comment]
There is one silly(?) issue with using `np.asarray(nan, dtype=old_out.dytpe) in the `switch`, when there are multiple chained log-exp ops AND using a float precision different than `config.floatX`.

```python
aesara.config.floatX = 'float32'
x = at.fvector('x')
z = at.exp(at.log(at.exp(at.log(x))))
f = aesara.function([x], z) # fine

x = at.vector('x')
z = at.exp(at.log(at.exp(at.log(x))))
f = aesara.function([x], z) # fine as x was converted to 32bit

aesara.config.floatX = 'float32'
x = at.fvector('x')
z = at.exp(at.log(at.exp(at.log(x))))
f = aesara.function([x], z) # fine

x = at.vector('x')  # x is converted to float32
z = at.exp(at.log(at.exp(at.log(x))))
f = aesara.function([x], z)  # fine

aesara.config.floatX = 'float64'
x = at.vector('x')
z = at.exp(at.log(at.exp(at.log(x))))
f = aesara.function([x], z)  # fine


aesara.config.floatX = 'float64'
x = at.fvector('x')
z = at.exp(at.log(at.exp(at.log(x))))
f = aesara.function([x], z)  # raises TypeError
```

```
ERROR (aesara.graph.opt): SeqOptimizer apply <aesara.tensor.basic_opt.FusionOptimizer object at 0x7f268e0a2730>
ERROR (aesara.graph.opt): Traceback:
ERROR (aesara.graph.opt): Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/aesara/aesara/scalar/basic.py", line 367, in filter
    raise TypeError(
TypeError: Value cannot accurately be converted to dtype (float32) and allow_downcast is not True
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/aesara/aesara/graph/opt.py", line 246, in apply
    sub_prof = optimizer.optimize(fgraph)
  File "/home/ricardo/Documents/Projects/aesara/aesara/graph/opt.py", line 84, in optimize
    ret = self.apply(fgraph, *args, **kwargs)
  File "/home/ricardo/Documents/Projects/aesara/aesara/tensor/basic_opt.py", line 4687, in apply
    new_outputs = self.optimizer(fgraph, node)
  File "/home/ricardo/Documents/Projects/aesara/aesara/tensor/basic_opt.py", line 4504, in local_fuse
    tmp.tag.test_value = tv.flatten()[0]
  File "/home/ricardo/Documents/Projects/aesara/aesara/graph/utils.py", line 267, in __setattr__
    obj = self.attr_filter(obj)
  File "/home/ricardo/Documents/Projects/aesara/aesara/scalar/basic.py", line 372, in filter
    raise TypeError(
TypeError: Could not convert <class 'numpy.float32'> (value=nan) to float32
```

The `FusionOptimizer` eventuall calls `Scalar(CType).filter` which fails these two checks `self.dtype == config.floatX ("float32" == "float64")` OR. `data == converted_data (nan == nan)`...

**Comment by brandonwillard** [comment]
> Would it make trouble to check instead whether `np.dtype(self.dtype) <= np.dtype(config.floatX)`?

Very interesting find!  I don't know; I'll have to take a look.

**Comment by brandonwillard** [comment]
It looks like we need another condition that uses `np.isnan` in `Scalar.filter`.

**Comment by brandonwillard** [comment]
I add a fix here: https://github.com/pymc-devs/aesara/pull/374.

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**